server/src/main/java/com/bfwg/rest/UserController.java

```diff
@@ -1,62 +1,86 @@
-
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-
-import static org.springframework.web.bind.annotation.RequestMethod.GET;
+import org.springframework.web.util.UriComponentsBuilder;
+import com.bfwg.exception.ResourceConflictException;
+import com.bfwg.model.Authority;
+import com.bfwg.model.User;
+import com.bfwg.service.AuthorityService;
+import com.bfwg.service.UserService;
 
 /**
  * Created by fan.jin on 2016-10-15.
  */
 
 @RestController
-@RequestMapping( value = "/api", produces = MediaType.APPLICATION_JSON_VALUE )
+@RequestMapping(value = "/api", produces = MediaType.APPLICATION_JSON_VALUE)
 public class UserController {
 
-    @Autowired
-    private UserService userService;
+  @Autowired
+  private UserService userService;
 
-    @RequestMapping( method = GET, value = "/user/{userId}" )
-    public User loadById( @PathVariable Long userId ) {
-        return this.userService.findById( userId );
-    }
+  @Autowired
+  private AuthorityService authService;
 
-    @RequestMapping( method = GET, value= "/user/all")
-    public List<User> loadAll() {
-        return this.userService.findAll();
-    }
+  @RequestMapping(method = GET, value = "/user/{userId}")
+  public User loadById(@PathVariable Long userId) {
+    return this.userService.findById(userId);
+  }
 
-    @RequestMapping( method = GET, value= "/user/reset-credentials")
-    public ResponseEntity<Map> resetCredentials() {
-        this.userService.resetCredentials();
-        Map<String, String> result = new HashMap<>();
-        result.put( "result", "success" );
-        return ResponseEntity.accepted().body(result);
-    }
-    /*
-     *  We are not using userService.findByUsername here(we could),
-     *  so it is good that we are making sure that the user has role "ROLE_USER"
-     *  to access this endpoint.
-     */
-    @RequestMapping("/whoami")
-    @PreAuthorize("hasRole('USER')")
-    public User user() {
-        return (User)SecurityContextHolder
-                .getContext()
-                .getAuthentication()
-                .getPrincipal();
+  @RequestMapping(method = GET, value = "/user/all")
+  public List<User> loadAll() {
+    return this.userService.findAll();
+  }
+
+  @RequestMapping(method = GET, value = "/user/reset-credentials")
+  public ResponseEntity<Map> resetCredentials() {
+    this.userService.resetCredentials();
+    Map<String, String> result = new HashMap<>();
+    result.put("result", "success");
+    return ResponseEntity.accepted().body(result);
+  }
+
+
+  @RequestMapping(method = POST, value = "/signup")
+  public ResponseEntity<?> addUser(@RequestBody User user, UriComponentsBuilder ucBuilder) {
+
+    User existUser = this.userService.findByUsername(user.getUsername());
+    if (existUser != null) {
+      throw new ResourceConflictException(user.getId(), "Username already exists");
     }
+    List<Authority> auths = this.authService.findByname("ROLE_USER");
```

Comment:
Can we relocate the below code into a method in a server? Possibly the user service?

Reply:
OK, will try tomorrow